src/main/scala/org/globalforestwatch/summarystats/forest_change_diagnostic/ForestChangeDiagnosticAnalysis.scala

```diff
@@ -0,0 +1,510 @@
+                                                points: util.HashSet[GeoSparkGeometry]
+                                              ): (FeatureId, ForestChangeDiagnosticDataLossYearly) = {
+    ( {
+      val id = {
+        poly.getUserData.asInstanceOf[String].filterNot("[]".toSet).toInt
+
+      }
+      FeatureIdFactory(featureType).featureId(id)
+
+    }, {
+      val fireCount =
+        points.asScala.toList.foldLeft(SortedMap[Int, Double]()) {
+          (z: SortedMap[Int, Double], point) => {
+            // extract year from acq_date column
+            val year = point.getUserData
+              .asInstanceOf[String]
+              .split("\t")(2)
+              .substring(0, 4)
+              .toInt
+            val count = z.getOrElse(year, 0.0) + 1.0
+            z.updated(year, count)
+          }
+        }
+
+      ForestChangeDiagnosticDataLossYearly.prefilled
+        .merge(ForestChangeDiagnosticDataLossYearly(fireCount))
+    })
+  }
+
+  private def reformatSummaryData(
+                                   summaryRDD: RDD[(FeatureId, ForestChangeDiagnosticSummary)]
+                                 ): RDD[(FeatureId, ForestChangeDiagnosticData)] = {
+
+    summaryRDD
+      .flatMap {
+        case (id, summary) =>
+          // We need to convert the Map to a List in order to correctly flatmap the data
+          summary.stats.toList.map {
+            case (dataGroup, data) =>
+              id match {
+                case featureId: SimpleFeatureId =>
+                  toForestChangeDiagnosticData(featureId, dataGroup, data)
+                case _ =>
+                  throw new IllegalArgumentException("Not a SimpleFeatureId")
+              }
+          }
+      }
+
+  }
+
+  private def aggregateFireData(
+                                 fires: ForestChangeDiagnosticDataLossYearly
+                               ): ForestChangeDiagnosticDataLossYearly = {
+    val minLossYear = fires.value.keysIterator.min
+    val maxLossYear = fires.value.keysIterator.max
+    val years: List[Int] = List.range(minLossYear + 1, maxLossYear + 1)
+
+    ForestChangeDiagnosticDataLossYearly(
+      SortedMap(
+        years.map(
+          year =>
+            (year, { // I hence declare them here explicitly to help him out.
+              val thisYearLoss: Double = fires.value.getOrElse(year, 0)
+              val lastYearLoss: Double = fires.value.getOrElse(year - 1, 0)
+              (thisYearLoss + lastYearLoss) / 2
+            })
+        ): _*
+      )
+    )
+  }
+
+  private def toForestChangeDiagnosticData(
+                                            featureId: FeatureId,
+                                            dataGroup: ForestChangeDiagnosticRawDataGroup,
+                                            data: ForestChangeDiagnosticRawData
+                                          ): (FeatureId, ForestChangeDiagnosticData) = {
+    (
+      featureId,
+      ForestChangeDiagnosticData(
```

Comment:
I see what you mean about better type checking by using all RDDs. However, I do find it's much less readable than using dataframes. It also has the same problem of just too much repetitive code, since we have to manually specify every single field and how to treat it. 

This is just becoming increasingly painful, since what should be small changes to the code can sometimes end up needing us to change the same thing 25-50 times for each of our 7 analyses. Somehow we need to parameterize our processing of these fields so we can this all in a single function mapping. I'm just still pretty novice at Scala, so I'm not sure of the best way, I'll mull it over for a bit.

Reply:
One possible way is we abstract "Field" to a case class, which would have like data type, layer name, output column name, etc., and then we define how to combine fields of the same type. Then we would just need a list of Fields to run our polygonal summary, which can then convert to a DF using the column names on the Field object. If we combine that with MultiBandTileArrays, perhaps we wouldn't need any of these explicit case classes.